src/dot-protocol/merlin_dot_protocol.mli

```diff
@@ -74,19 +74,53 @@ end
 
 type directive = Directive.Processed.t
 
-module Commands : sig
-  type t = File of string | Halt | Unknown
-
-  val read_input : in_channel -> t
-  val send_file : out_channel:out_channel -> string -> unit
-end
-
 type read_error =
   | Unexpected_output of string
   | Csexp_parse_error of string
 
-(** [read inc] reads one csexp from the channel [inc] and returns the list of
-  directives it represents *)
-val read : in_channel:in_channel -> (directive list, read_error) Merlin_utils.Std.Result.t
+module Make (IO : sig
+  type 'a t
+
+  module O : sig
+    val ( let+ ) : 'a t -> ('a -> 'b) -> 'b t
+  end
+end) (Chan : sig
+  type in_chan
+  type out_chan
+
+  val read : in_chan -> (Csexp.t, string) result IO.t
+
+  val write : out_chan -> Csexp.t -> unit IO.t
+end) : sig
+  (** [read inc] reads one csexp from the channel [inc] and returns the list of
+      directives it represents *)
+  val read : Chan.in_chan -> (directive list, read_error) Merlin_utils.Std.Result.t IO.t
+
+  val write : Chan.out_chan -> directive list -> unit IO.t
+
+  module Commands : sig
+    type t = File of string | Halt | Unknown
+    val read_input : Chan.in_chan -> t IO.t
+
+    val send_file : Chan.out_chan -> string -> unit IO.t
 
-val write : out_channel:out_channel -> directive list -> unit
+    val halt : Chan.out_chan -> unit IO.t
+  end
+end
+
+module Std : sig
+  module Commands :
+  sig
```

Comment:
We could name that sig to avoid duplicating it (module type S = sig end)

Reply:
Forgot about this one: the signature of the `Blocking` module is, as expected, the same as the signature of the body of the functor Make.

You can give a name to that signature and use it both for Make and Blocking